Dem dmt force model (#1045)
lethe-cfd/lethe: doc/source/theory/dem/dem.rst

    \delta_n =& \:R_i + R_j - |\mathbf{x}_{j} - \mathbf{x}_{i}| \\
    \mathbf{\delta}_{ij}^{t,\text{new}} &= \mathbf{\delta}_{ij}^{t,\text{old}}+\mathbf{v}_{rt}dt

~~~~~~~~~~~~~~~~~~~~~

> Relative velocities

Reviewer: I suggest you use the Title case capitalization (words are capitalized, except for minor words) as done for other subsections in documentation.

> ~~~~~~~~~~~~~~~~~~~~~

Reviewer: Can you remove the warning "This page is under construction" at the top? This page is now quite detailed enough to not still be under construction :p

> The relative velocities are calculated to update the tangential overlap :

Reviewer: ```suggestion
The relative velocities are calculated to update the tangential overlap:
```

.. math::
    \mathbf{v}_{ij} &= \mathbf{v}_i-\mathbf{v}_j+\left(R_i\mathbf{\omega}_i+R_j\mathbf{\omega}_j\right)\times\mathbf{n}_{ij}
    \mathbf{v}_{rn} &= \left(\mathbf{v}_{ij}.\mathbf{n}_{ij}\right)\mathbf{n}_{ij} \\
    \mathbf{v}_{rt} &= \mathbf{v}_{ij}-\mathbf{v}_{rn} \\

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Spring and damping constants
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The spring and damping constants for the linear and nonlinear viscoelastic models are calculated as follow:

Reviewer 2: ```suggestion
The spring and damping constants for the linear and nonlinear viscoelastic models are calculated as follows:
```

> .. list-table:: Spring and Damping Models used in Lethe.

Reviewer 2: ```suggestion
.. list-table:: Spring and damping models used in Lethe.
```

* :math:`\nu_i, \nu_j` poisson coefficient of particle i or j;

~~~~~~~~~~~~~~~~~~~~
Coulomb's limit
~~~~~~~~~~~~~~~~~~~~

Coulomb's criterion is breached when the following condition is broken during a collision:

.. math::
    |\mathbf{F}_{ij}^{t}| \geq \mu |\mathbf{F}_{ij}^{n}|

> A breach means the collision is having gross sliding and tangential force needs to be limited to the Coulomb's limit.

Author: Yes, but what does it mean? Your version is only saying what is happening algorithmically, not physically...

Also, like I said in a previous comment to Bruno, my next PR is about the Coulomb limit, so I wont focus on modifying this in the documentation since it will change in the next two weeks.

.. math::
    \mathbf{\delta_t} &= \frac{\mathbf{\tilde{F}_{ij}}}{-k_{t}} \\
    \mathbf{\tilde{F}_{ij}} &= \mathbf{\hat{F}_{ij}} + \eta_{t}\mathbf{v}_{rt} \\
    \mathbf{\hat{F}_{ij}^{t}} &= \mu |\mathbf{F}_{ij}^{n}| \frac{\mathbf{F}_{ij}^{t}}{|\mathbf{F}_{ij}^{t}|}

Regarding the particle-wall contacts, the models applied are the same as particle-particle contacts with a background triangulation and mapping with walls.

.. note::
    When using a cohesive force models, Coulomb's criterion need to be modified. For further information, please read the sections associated with each cohesive force model.

~~~~~~~~~~~~~~~~~~~~~~~~~
Tangential torque
~~~~~~~~~~~~~~~~~~~~~~~~~

Tangential torque is the torque generated by the tangential force. It can be calculated through:

.. math::
    \mathbf{M}_{ij}^{t} = R_{i}\mathbf{n}_{ij} \times \mathbf{F}_{ij}^{t}

.. note::
    As of now, the lethe-particles solver only uses spherical particles, thus the normal force does not generate a torque on the particle during a collision.

~~~~~~~~~~~~~~~~~~~~~~~~~
Rolling friction models
~~~~~~~~~~~~~~~~~~~~~~~~~

> Rolling friction may be computed through a constant torque model or a viscous torque model. It can also be disabled using a no resistance torque model. The corresponding model can be described by the following equations:

Reviewer 2: Good question, I'm not sure if we have an explicit no model

> Rolling friction may be computed through a constant torque model or a viscous torque model. It can also be disabled using a no resistance torque model. The corresponding model can be described by the following equations:

Author: I get the confusion, but I still thing it is relevant to have it in the table.

> Rolling friction may be computed through a constant torque model or a viscous torque model. It can also be disabled using a no resistance torque model. The corresponding model can be described by the following equations:

Author: it is implemented in lethe: 

https://lethe-cfd.github.io/lethe/documentation/parameters/dem/model_parameters.html

"no_resistance"

> Rolling friction may be computed through a constant torque model or a viscous torque model. It can also be disabled using a no resistance torque model. The corresponding model can be described by the following equations:

Author: We are not talking about parameters on this page. Only theory. 

This would be inconsistent with the rest of this page. (ex. "young modulus particles" , "poisson ratio particles", etc.)

Also every contact force model.

> Rolling friction may be computed through a constant torque model or a viscous torque model. It can also be disabled using a no resistance torque model. The corresponding model can be described by the following equations:

Reviewer: In the table yes but in the text "using a no resistance torque model" sounds weird no?

.. list-table:: Rolling Friction Models used in Lethe.
   :widths: 30 30 30
   :header-rows: 1
   :align: center

   * - No resistance
     - Constant resistance
     - Viscous resistance
   * - :math:`\mathbf{M}_{ij}^{r} = 0`
     - :math:`\mathbf{M}_{ij}^{r} = -\mu_{r}R_{e}|\mathbf{F}_{ij}^{n}| \mathbf{\hat{\omega}}_{ij}`
     - :math:`\mathbf{M}_{ij}^{r} = -\mu_{r}R_{e}|\mathbf{F}_{ij}^{n}||\mathbf{V}_{\omega}| \mathbf{\hat{\omega}}_{ij}`

Where:

* :math:`\mu_{r}` rolling friction coefficient;
* :math:`\hat{\omega}_{ij}` relative angular velocity;
* :math:`V_{\omega}` contact point relative velocity caused by the angular velocities;   // This need to be checked

> The parameters are computed as followed:

Author: Good question! I have stayed conherent with the term we use for the effective mass, effective radius, etc. 

I can change it if you have a better term, but I think it's fine.

.. math::
    \mathbf{\hat{\omega}}_{ij} &= \frac{\omega_{i} - \omega_{j}}{|\omega_{i} - \omega_{j}|} \\
    \mathbf{V}_{\omega} &= \left( \omega_{i} \times R_{i}\mathbf{n}_{ij}-\omega_{j} \times R_{j}\mathbf{n}_{ji} \right).

-----------------------
Cohesive force models
-----------------------

> Lethe supports two cohesive force models: the Johnson-Kendall-Roberts (JKR) and the Derjaguin-Muller-Toporov (DMT). Both models describe attractive forces due to van der Waals effects. Choosing the right model can be based on the Tabor parameter :math:`\mathbf{\tau}` which represents the ratio between the normal elastic deformation caused by adhesion and the distance at which adhesion forces occur. `[3] <https://doi.org/10.1163/1568561054352685>`_

Author: It's not a number, it's a parameter.

This parameter can be described as:

.. math::
    \mathbf{\tau} = \left( \frac{R_{e} \gamma_{e}^2}{Y_{e}^2 z_{o}^3}\right)^{1/3}.

Where :math:`\mathbf{z_{o}}` is the equilibrium separation of the surfaces and :math:`\mathbf{\gamma}_{e}` the effective surface energy. The DMT model is applicable for low :math:`\mathbf{\tau}` values (:math:`\mathbf{\tau} < 1`) while the JKR model is more appropriate for high :math:`\mathbf{\tau}` values (:math:`\mathbf{\tau} > 1`) . In essence, the DMT model is preferred for small, hard particles (high :math:`Y`) and the JKR model for large, soft particles.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Johnson-Kendall-Roberts force model
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The Johnson-Kendall-Roberts (JKR) model describes attractive forces due to van der Waals effects. `[4] <https://doi.org/10.3390/pr11010005>`_
This model modifies the Hertz formulation by defining a larger contact path radius (:math:`\mathbf{a}`) and by taking into account the effective surface energy (:math:`\mathbf{\gamma}_{e}`).
The model is defined by:

.. math::

>     a^{3} = \frac{3 R_{e}}{4 Y_{e}} \left[F_{n}^{JKR} + 3\pi\gamma_{e}R_{e}  + \sqrt{6 F_{n}^{JKR} \pi\gamma_{e}R_{e} + (3\pi\gamma_{e}R_{e})^2 }\right].

Reviewer: Fn is a vector (bold)

Where :math:`\mathbf{F_{n}^{JKR}}` corresponds to the normal spring force and attractive force combined and :math:`\mathbf{\gamma_{e}}` is the effective surface energy.
Note that if the effective surface energy is equal to zero, the JKR model reverts to Hertz model.

The effective surface energy can be computed as:

.. math::
    \gamma_{e} = \gamma_{i} + \gamma_{j} - 2\gamma_{i,j}.

Where :math:`\gamma_{i}` and :math:`\gamma_{j}` are the surface energy of each material (particle or wall) and where :math:`\gamma_{i,j}` is the interface energy which is equal to zero when both surfaces are the same material.
In Lethe, the interface energy term is approximated using `[5] <https://doi.org/10.1016/B978-0-12-391927-4.10013-1>`_:

.. math::
    \gamma_{i,j} \approx \left( \sqrt{\gamma_{i}} - \sqrt{\gamma_{j}}  \right)^{2}.

> 

Author: i and j are well defined at tthe start of the theory.

To compute the :math:`\mathbf{F_{n}^{JKR}}`, the contact patch radius needs to be determined. The contact patch radius can be related to the normal overlap as follows:

.. math::
    \delta_{n} = \frac{ a^{2} }{ R_{e} } -  \sqrt{ \frac{2 \pi \gamma_{e} a }{ Y_{e} }}.

This equation can be rewritten as a fourth-order polynomial function with two complex and two real roots.

.. math::
    0 = a^{4} - 2R_{e}\delta_{n}a^{2} - 2\pi\gamma_{e}R_{e}^{2}a + R_{e}^{2}\delta_{n}^{2}

> Since we are always solving for the same real root, a straightforward procedure, described by Parteli et al. can be used `[6] <https://doi.org/10.1038/srep06227>`_:

Reviewer: It's 
c_{1} &= -2\pi\gamma_{e}R_{e}^{2}/Y_{e}
in the code

Author: Good catch

* :math:`\nu_i, \nu_j` poisson coefficient of particle i or j;

~~~~~~~~~~~~~~~~~~~~
Coulomb's limit
~~~~~~~~~~~~~~~~~~~~

> Coulomb's criterion is violated when the following condition is not respected during a collision:

Reviewer: ```suggestion
Coulomb's criterion is violated when the following condition is broken during a collision:
```

.. math::
    |\mathbf{F}_{ij}^{t}| \geq \mu |\mathbf{F}_{ij}^{n}|

> A violation means the collision is having gross sliding and tangential force needs to be limited to the Coulomb limit.

Reviewer: Coulomb limit or Coulomb's limit, by consistent

To do so, the tangential overlap :math:`\mathbf{\delta_t}` is first limited and then the tangential force is recalculated.

> The tangential overlap is calculated with the tangential force with no damping force as default nonlinear contact model as follow:

Author: I have an other PR coming about the tangential overlap and the Coulomb  limit. Do you think this could wait for the PR? 

It's weird syntax and it will become change in the code...

> The tangential overlap is calculated with the tangential force with no damping force as default nonlinear contact model as follow:

Reviewer 2: okidoke

.. math::
    \mathbf{\delta_t} &= \frac{\mathbf{\tilde{F}_{ij}}}{-k_{t}} \\
    \mathbf{\tilde{F}_{ij}} &= \mathbf{\hat{F}_{ij}} + \eta_{t}\mathbf{v}_{rt} \\
    \mathbf{\hat{F}_{ij}^{t}} &= \mu |\mathbf{F}_{ij}^{n}| \frac{\mathbf{F}_{ij}^{t}}{|\mathbf{F}_{ij}^{t}|}

> Regarding the particle-wall contacts, applied models are the same than particle-particle contacts with a background triangulation and mapping with walls.

Author: I think it was about the solid-object getting map in the triangulation, but like I told you I just reorganized the theory, so I'm not actually the one who wrote this...

> Regarding the particle-wall contacts, applied models are the same than particle-particle contacts with a background triangulation and mapping with walls.

Reviewer: Hehe

.. note::

>     When using a cohesive force models, Coulomb's criterion need to be adjusted, since the normal contact force isn't equal to the total normal force between particle in a collision.

Reviewer: You can give the new condition as an equation instead of a note if you want

Author: it depends on the cohesive force model... also it's still something I want to make sure I'm doing correctly in the code

Reviewer: Ok, once it's done correctly then haha

Author: it's done!

>     When using a cohesive force models, Coulomb's criterion need to be adjusted, since the normal contact force isn't equal to the total normal force between particle in a collision.

Author: We only use the normal spring + normal damping force. 

We don't use the cohesive force. There's an article I can give you which is on this topic.

>     When using a cohesive force models, Coulomb's criterion need to be adjusted, since the normal contact force isn't equal to the total normal force between particle in a collision.

Author: For the JKR model, I am following PFC implementation based on the work of Thornton (1991). (The article is on my Zotero if you want to check)

For the DMT model, since I'm not able to find a reference on how to implement Coulomb's limit, I will just ignore the cohesive term when computing the total "coulomb_threshold" variable.  Otherwise, particle can slip indefinitely which is not physically possible.

~~~~~~~~~~~~~~~~~~~~~~~~~
Tangential torque
~~~~~~~~~~~~~~~~~~~~~~~~~

Tangential torque is the torque generated by the tangential force. It can be calculated through:

.. math::
    \mathbf{M}_{ij}^{t} = R_{i}\mathbf{n}_{ij} \times \mathbf{F}_{ij}^{t}

.. note::

>     As of February 2024, the lethe-particles solver only uses spherical particles, thus the normal force does not generate a torque on the particle during a collision.

Reviewer: This is very specific, but that's fine haha

>     As of February 2024, the lethe-particles solver only uses spherical particles, thus the normal force does not generate a torque on the particle during a collision.

Reviewer 2: ```suggestion
    As of now, the lethe-particles solver only uses spherical particles, thus the normal force does not generate a torque on the particle during a collision.
```
We will change it when that changes :)

Author: Indeed, that's the plan! :))

~~~~~~~~~~~~~~~~~~~~~~~~~
Rolling friction models
~~~~~~~~~~~~~~~~~~~~~~~~~

> Rolling friction may be calculated through a constant torque model or a viscous torque model. The user also has the option of not using a rolling friction model. The corresponding model can be described by the following equations:

Reviewer: Since this is the Theory guide, I would not mention "the user" or options. This is more appropriate for the Parameters guide

.. list-table:: Rolling Friction Models used in Lethe.
   :widths: 30 30 30
   :header-rows: 1

   * - No resistance
     - Constant resistance
     - Viscous resistance
   * - :math:`\mathbf{M}_{ij}^{r} = 0`
     - :math:`\mathbf{M}_{ij}^{r} = -\mu_{r}R_{e}|\mathbf{F}_{ij}^{n}| \mathbf{\hat{\omega}}_{ij}`
     - :math:`\mathbf{M}_{ij}^{r} = -\mu_{r}R_{e}|\mathbf{F}_{ij}^{n}||\mathbf{V}_{\omega}| \mathbf{\hat{\omega}}_{ij}`

Where:

* :math:`\mu_{r}` rolling friction coefficient;
* :math:`\hat{\omega}_{ij}` relative angular velocity;

> * :math:`V_{\omega}` contact point reltative velocity caused by the angular velocities;   // This need to be checked

Author: Good catch!

-----------------------
Cohesive force models
-----------------------
Lethe supports two cohesive force models named the Johnson-Kendall-Roberts (JKR) and the Derjaguin-Muller-Toporov (DMT). Both models describes attractive forces due to van der Waals effects. Choosing the right model can be based on the Tabor parameter :math:`\mathbf{\tau}` which represents the ratio between the normal elastic deformation caused by adhesion and the distance at which adhesion forces occur. `[3] <https://doi.org/10.1163/1568561054352685>`_

This parameter can be described as:

.. math::

>     \mathbf{\tau} = \left( \frac{R_{e} \gamma_{e}^2}{E_{eq}^2 z_{o}^3}\right)^{1/3}.

Reviewer 2: Equivalent Young modulus I guess? Has it been defined?

>     \mathbf{\tau} = \left( \frac{R_{e} \gamma_{e}^2}{E_{eq}^2 z_{o}^3}\right)^{1/3}.

Author: It should be Y_e , not E_e

Y_e is defined

Where :math:`\mathbf{z_{o}}` is the equilibrium separation of the surfaces and :math:`\mathbf{\gamma}_{e}` the effective surface energy. For low :math:`\mathbf{\tau}` values (:math:`\mathbf{\tau} < 1` ) the DMT model is applicable where as for high :math:`\mathbf{\tau}` values (:math:`(\mathbf{\tau} > 1 )` the JKR model is more appropriate. In essence, the DMT model is preferred for small, hard particles (high :math:`\mathbf{Y}`) and the JKR model for large, soft particles.

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Johnson-Kendall-Roberts force model
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
The Johnson-Kendall-Roberts (JKR) model describes attractive forces due to van der Waals effects. `[4] <https://doi.org/10.3390/pr11010005>`_

> This model modifies the Hertz formulation by defining a larger contact path radius (:math:`\mathbf{a}`) and taking into account the effective surface energy (:math:`\mathbf{\gamma}_{e}`).

Author: `This model modifies the Hertz formulation by defining a larger contact path radius (:math:`\mathbf{a}`) and by taking into account the effective surface energy (:math:`\mathbf{\gamma}_{e}`).`

> This model modifies the Hertz formulation by defining a larger contact path radius (:math:`\mathbf{a}`) and taking into account the effective surface energy (:math:`\mathbf{\gamma}_{e}`).

Author: ```suggestion
This model modifies the Hertz formulation by defining a larger contact path radius (:math:`\mathbf{a}`) and by taking into account the effective surface energy (:math:`\mathbf{\gamma}_{e}`).
```

    \end{cases}\\
    \omega &= \sqrt{c_{2} + 2 s} \\
    \lambda &= \frac{c_{1} }{2 \omega}\\
    a &= \frac{1}{2}\left(\omega + \sqrt{\omega^{2} - 4(c_{2} + s + \lambda ) } \right).

Finally, the :math:`\mathbf{F_{n}^{JKR}}` can be computed as follows:

.. math::

>     F_{n}^{JKR} = \frac{4 Y_{e} a^{3}}{3 R_{e}} - \sqrt{8 \pi \gamma_{e} Y_{e} a^{3}}.

Reviewer: Check if all the vector are bold

The normal damping, tangential damping and tangential spring constants need to be computed using the same procedure as the nonlinear model.

A simplified version of the JKR model (SJKR-A) is implemented in Lethe. Please refer to C. J. Coetzee and O. C. Scheffler for more information on the different versions of the JKR model and their specific features. `[4] <https://doi.org/10.3390/pr11010005>`_

A modified Coulomb's limit, based on the work of C. Thornton `[7] <https://doi.org/10.1088/0022-3727/24/11/007>`_, is used for the JKR model. Using the usual limit can result in permanent slip since the total normal force can be equal to zero even when there is a substantial overlap between particles.

The modified Coulomb's criterion is breached when the following condition is broken during a collision:

.. math::
    |\mathbf{F}_{ij}^{t}| \geq \mu |\mathbf{F_{n}^{JKR} + 2F_{po}}|.

Where :math:`\mathbf{2F_{po}}` is the pull-off force, which can be computed as follows:

.. math::

>     \mathbf{F_{po}} = 1.5\pi\gamma_{e}R_{e}.

Reviewer: You use 3 in the code (with the 2* in the criterion calculation)

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
Derjaguin-Muller-Toporov force model
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The Derjaguin-Muller-Toporov (DMT) model describes attractive forces due to van der Waals effects. This model is more suitable for particles with smaller diameter, lower surface energy and higher Young's modulus. In lethe, the DMT model is implemented using the Maugis approximation which simply adds an adhesion term :math:`\mathbf{F_{ad}^{DMT}}` to the normal force calculation. `[8] <https://doi.org/10.1016/j.prostr.2018.11.106.>`_

Reviewer 3: I would mention again the objective criteria for this. "Smaller and harder particles" seems a bit vague and you don't need to read the full documentation to find this information if you bring it here.

.. math::
    F_{ad}^{DMT} = -2\pi\gamma_{e}R_{e}^{2}

> The Coulomb's limit threshold for the DMT model is computed in the same way as for the non-linear viscoelastic model. This means that the adhesion force term in not taken into account when computing the norm of the normal force.

Reviewer 3: Would you add a hyperlink for the section here?

The normal damping, tangential damping and tangential spring constants need to be computed using the same procedure as the nonlinear model.

> For implementation reasons, a simplified version of the JKR model (SJKR-A) is implemented in Lethe. This version does not calculate non-contact forces. Please refer to C. J. Coetzee and O. C. Scheffler for more information on the different versions of the JKR model and their specific features. `[4] <https://doi.org/10.3390/pr11010005>`_

Reviewer 2: ```suggestion
A simplified version of the JKR model (SJKR-A) is implemented in Lethe. This version does not calculate non-contact forces. Please refer to C. J. Coetzee and O. C. Scheffler for more information on the different versions of the JKR model and their specific features. `[4] <https://doi.org/10.3390/pr11010005>`_
```
It's not clear what you mean here by non-contact force.

Author: You are right. There's a nuance between non contact and (R_1 + R_2) > Distance. 

Clearly, I'll need to write about this in my litterature review. 

I think I will just state that the SJKR-A is implemented and refer to Coetzee article. Otherwise I need to write about the hysteresis of the Full-JKR model on this page and it's weird to describe in detail how a certain model works when this model isn't implemented in Lethe...

Reviewer: You could just explain in one sentence what does non-contact means here without refering to a model
